src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1279,7 +1280,42 @@ private void Abort(Exception abortException)
             return LifetimeExpired(nowTicks, connectionLifetime);
         }
 
-        private void AbortStreams(int lastValidStream, Exception abortException)
+        private void AbortStreams(Exception abortException)
+        {
+            lock (SyncObject)
+            {
+                if (NetEventSource.IsEnabled) Trace($"{nameof(abortException)}={abortException}, {nameof(_lastValidStreamId)}={_lastValidStreamId}");
+
+                bool hasAnyActiveStream = false;
+                foreach (KeyValuePair<int, Http2Stream> kvp in _httpStreams)
+                {
+                    int streamId = kvp.Key;
+                    Debug.Assert(streamId == kvp.Value.StreamId);
+
+                    kvp.Value.OnAbort(abortException);
+                    _httpStreams.Remove(streamId);
+                }
+
+                if (!hasAnyActiveStream)
+                {
+                    _disposed = true;
+                }
+
+                CheckForShutdown();
+            }
+
+            if (!_disposed)
+            {
+                _disposed = true;
+
+                // Invalidate outside of lock to avoid race with HttpPool Dispose()
+                // We should not try to grab pool lock while holding connection lock as on disposing pool,
+                // we could hold pool lock while trying to grab connection lock in Dispose().
+                _pool.InvalidateHttp2Connection(this);
```

Comment:
I wonder if it's better to invalidate the connection before killing the streams, just to try to better avoid any new requests trying to use the connection.

Note it's fine to call InvalidateHttp2Connection multiple times on the same connection; we check in InvalidateHttp2Connection whether the connection is actually the one we're using currently or not.

Same comment for StartTerminatingConnection below.

Reply:
we currently process frames synchronously so I'm not sure it will make a difference but will change it